include/fuzzuf/algorithms/nautilus/grammartec/tree.hpp

```diff
@@ -0,0 +1,174 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+/**
+ * @file tree.hpp
+ * @brief Tree for context-free grammar
+ * @author Ricerca Security <fuzzuf-dev@ricsec.co.jp>
+ */
+#ifndef FUZZUF_INCLUDE_ALGORITHMS_NAUTILUS_GRAMMARTEC_TREE_HPP
+#define FUZZUF_INCLUDE_ALGORITHMS_NAUTILUS_GRAMMARTEC_TREE_HPP
+
+#include <optional>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <utility>
+#include <variant>
+#include <vector>
+#include "fuzzuf/algorithms/nautilus/grammartec/context.hpp"
+#include "fuzzuf/algorithms/nautilus/grammartec/newtypes.hpp"
+#include "fuzzuf/algorithms/nautilus/grammartec/rule.hpp"
+
+
+namespace fuzzuf::algorithm::nautilus::grammartec {
+
+class RecursionInfo;
+class Tree;
+class TreeLike;
+class TreeMutation;
+class Unparser;
+class UnparserStep;
+
+
+class TreeLike {
+public:
+  virtual const RuleID& GetRuleID(const NodeID& n) const = 0;
+  virtual size_t Size() const = 0;
+  virtual Tree ToTree(Context& ctx) const = 0;
+  virtual const Rule& GetRule(const NodeID& n, Context& ctx) const = 0;
+  virtual const RuleIDOrCustom& GetRuleOrCustom(const NodeID& n) const = 0;
+  virtual const std::string& GetCustomRuleData(const NodeID& n) const = 0;
+  const NTermID& GetNontermID(const NodeID& n, Context& ctx) const;
+
+  void Unparse(const NodeID& id, Context& ctx, std::string& data) const;
+  void UnparseTo(Context& ctx, std::string& data) const;
+  std::string UnparseNodeToVec(const NodeID& n, Context& ctx) const;
+  std::string UnparseToVec(Context& ctx) const;
+};
+
+
+class Tree: public TreeLike {
+public:
+  Tree(std::vector<RuleIDOrCustom>&& rules, Context& ctx);
+  Tree(std::vector<RuleIDOrCustom>&& rules,
+       std::vector<size_t>&& sizes,
+       std::vector<NodeID>&& paren)
+    : _rules(std::move(rules)),
+      _sizes(std::move(sizes)),
+      _paren(std::move(paren)) {};
+  Tree(const std::vector<RuleIDOrCustom>& rules, // constructor with copy
+       const std::vector<size_t>& sizes,
+       const std::vector<NodeID>& paren)
+    : _rules(rules), _sizes(sizes), _paren(paren) {};
+  std::vector<RuleIDOrCustom>& rules() { return _rules; }
+  std::vector<size_t>& sizes() { return _sizes; }
+  std::vector<NodeID>& paren() { return _paren; }
+
+  virtual const RuleID& GetRuleID(const NodeID& n) const;
+  virtual size_t Size() const;
+  virtual Tree ToTree(Context& ctx) const;
+  virtual const Rule& GetRule(const NodeID& n, Context& ctx) const;
+  virtual const RuleIDOrCustom& GetRuleOrCustom(const NodeID& n) const;
+  virtual const std::string& GetCustomRuleData(const NodeID& n) const;
```

Comment:
Is `Tree` meant to be used as a base class?  Otherwise virtual specifiers are not needed. Probably you can specify override (or final) instead.

Reply:
Fixed in [29474e7](https://github.com/fuzzuf/fuzzuf/pull/23/commits/29474e7f48c06242224c5734dd0d3c70030817dc)